### ISSUE TITLE
TCP RST was sent with large response body

### ISSUE THREAD

**Issue by keimoon** [akka/akka-http#459]
Akka HTTP version: 3.0.0-RC1

How to reproduce:

 - Use small MTU for network interface (for example 1500). You can set MTU for loopback interface with `sudo ifconfig lo0 mtu 1500`
 - Response with large body content (should be larger than 128KiB) using `complete` and `ToResponseMarshallable`
 - Make sure client send header `Connection: close`.

Expected result:

Client should receive normally

Actual result:

Client got `Recv failure: Connection reset by peer`

This causes problems with some proxy like nginx or kong because they send `Connection: close` to upstream by default.

**Comment by ktoso** [comment]
Thanks for reporting. I'll mark as bug though I'm not sure about it.

**Comment by jrudolph** [comment]
It could be like this: when the server sees `Connection: close` on a request it fully closes the connection directly after having sent out the last bit of the response which might close the socket too soon?

 @keimoon which client did you use? Could you try to capture a network trace using tcpdump and attach it here? Thanks ;)

**Comment by keimoon** [comment]
I used curl. Here is network trace that you can load into wireshark


[akka.pcap.tar.gz](https://github.com/akka/akka-http/files/566522/akka.pcap.tar.gz)

![screen shot 2016-11-02 at 7 34 33 pm](https://cloud.githubusercontent.com/assets/1174062/19928936/74234002-a133-11e6-8cbc-97b5a3d92df7.png)

**Comment by keimoon** [comment]
Also I can only reproduce the bug with HttpEntity.Strict

**Comment by jrudolph** [comment]
Great, thanks @keimoon. That will help a lot.

**Comment by jrudolph** [comment]
I had a quick try, this issue can be reproduced using your instructions. 

The problem is that when the HTTP server decides to close the connection, it does so by completely killing the HTTP stack which will complete the write-side of the TCP connection stream and cancel the read-side. If cancel arrives first (which seems to be the case), the Tcp stream implementation will abort the connection (= socket.close with soLinger(0)). The result is that the OS just doesn't know about the connection any more and if any data was not sent yet this data is lost and any further packets will be responded with RST.

We've seen issues like that for a long time (IIRC also with spray). I guess the solution could be just to keep cancellation from reaching the TCP connection. (I don't think that this creates leaks, hopefully leaks would be prevented by the idle-timeout on the connection but it might make sense to think a bit about it.)

**Comment by keimoon** [comment]
Great. Currently we are implementing an work around that will use HttpEntity.Chunked if the response body is large.

**Comment by mariszin** [comment]
We migrated our app to akka-http 10.0.0, and received the same error - when using anything, that sends header "Connection: close" (nginx, curl), the request will fail on Strict entities, that are larger than 128K.
Error in curl:
curl: (56) Recv failure: Connection reset by peer

**Comment by guntiso** [comment]
Chunked does not help - changed response entity to HttpEntity.Chunked(mediaType, Source(Seq(ByteString(prettyJsonString)))), and now it works with curl, but still fails with nginx. Is there any reliable workaround?

**Comment by daandi** [comment]
Same here :( Would really appreciate a way to solve this.

**Comment by ktoso** [comment]
So if I understand correctly, we'd need a "CancellationBarrier" that can be "opened" by external signal, which we could use to resolve the race (making it always "cancel" after the completion went "all the way"). Reasonable @jrudolph ?

**Comment by jrudolph** [comment]
We might not need to send cancellation to the tcp stream at all but I haven't checked if that would work.

**Comment by jrudolph** [comment]
I reproduced it with this code https://github.com/jrudolph/akka-http/commit/a929a03f9d68177f9f04153dac6409acddbcc9dd and curl:

```
curl -vvv -H "connection:close" http://localhost:9001/big > /dev/null
```

**Comment by jonas** [comment]
Tested a locally published version of Akka HTTP and I no longer see connection resets.

**Comment by ktoso** [comment]
Cool, thanks for testing!

**Comment by antonkatz** [comment]
I am perhaps seeing this again with nginx.
I'm serving an html page with Akka-Http 10.0.9 with `getFromFile()`.
The file contains an <img> tag with `src` containing base64 encoded image.

Nginx has this in the error logs.
`upstream prematurely closed connection while reading upstream`

The error is resolved by using the workaround in akka/akka#19542

**Comment by jonas** [comment]
@antonkatz Is the traffic served over HTTPS?

**Comment by jrudolph** [comment]
@antonkatz we fixed issue #1219 in 10.0.10 which was similar to this bug when running on HTTPS. Can you check if 10.0.10 still has the issue?
